File: aegis/backend/test/internal/utils/data_transformation_test.go
package utils_test

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"

	"github.com/balkanid/aegis-backend/internal/utils"
)

func TestFormatFileSize(t *testing.T) {
	tests := []struct {
		bytes    int64
		expected string
	}{
		{0, "0 Bytes"},
		{512, "512.00 Bytes"},
		{1024, "1.00 KB"},
		{1536, "1.50 KB"},
		{1048576, "1.00 MB"},
		{1073741824, "1.00 GB"},
		{1099511627776, "1.00 TB"},
		{1125899906842624, "1024.00 TB"}, // Large number
	}

	for _, test := range tests {
		result := utils.FormatFileSize(test.bytes)
		assert.Equal(t, test.expected, result, "FormatFileSize(%d) = %s, expected %s", test.bytes, result, test.expected)
	}
}

func TestGetMimeTypeFromExtension(t *testing.T) {
	tests := []struct {
		filename string
		expected string
	}{
		{"test.txt", "text/plain"},
		{"test.html", "text/html"},
		{"test.css", "text/css"},
		{"test.js", "application/javascript"},
		{"test.json", "application/json"},
		{"test.pdf", "application/pdf"},
		{"test.jpg", "image/jpeg"},
		{"test.jpeg", "image/jpeg"},
		{"test.png", "image/png"},
		{"test.mp4", "video/mp4"},
		{"test.mp3", "audio/mpeg"},
		{"test.zip", "application/zip"},
		{"test.unknown", "application/octet-stream"},
		{"noextension", "application/octet-stream"},
		{"", "application/octet-stream"},
	}

	for _, test := range tests {
		result := utils.GetMimeTypeFromExtension(test.filename)
		assert.Equal(t, test.expected, result, "GetMimeTypeFromExtension(%s) = %s, expected %s", test.filename, result, test.expected)
	}
}

func TestIsImageMimeType(t *testing.T) {
	tests := []struct {
		mimeType string
		expected bool
	}{
		{"image/jpeg", true},
		{"image/png", true},
		{"image/gif", true},
		{"text/plain", false},
		{"video/mp4", false},
		{"", false},
	}

	for _, test := range tests {
		result := utils.IsImageMimeType(test.mimeType)
		assert.Equal(t, test.expected, result, "IsImageMimeType(%s) = %v, expected %v", test.mimeType, result, test.expected)
	}
}

func TestIsVideoMimeType(t *testing.T) {
	tests := []struct {
		mimeType string
		expected bool
	}{
		{"video/mp4", true},
		{"video/avi", true},
		{"image/jpeg", false},
		{"text/plain", false},
		{"", false},
	}

	for _, test := range tests {
		result := utils.IsVideoMimeType(test.mimeType)
		assert.Equal(t, test.expected, result, "IsVideoMimeType(%s) = %v, expected %v", test.mimeType, result, test.expected)
	}
}

func TestIsAudioMimeType(t *testing.T) {
	tests := []struct {
		mimeType string
		expected bool
	}{
		{"audio/mpeg", true},
		{"audio/wav", true},
		{"image/jpeg", false},
		{"text/plain", false},
		{"", false},
	}

	for _, test := range tests {
		result := utils.IsAudioMimeType(test.mimeType)
		assert.Equal(t, test.expected, result, "IsAudioMimeType(%s) = %v, expected %v", test.mimeType, result, test.expected)
	}
}

func TestIsArchiveMimeType(t *testing.T) {
	tests := []struct {
		mimeType string
		expected bool
	}{
		{"application/zip", true},
		{"application/x-rar-compressed", true},
		{"application/x-7z-compressed", true},
		{"application/gzip", true},
		{"text/plain", false},
		{"image/jpeg", false},
		{"", false},
	}

	for _, test := range tests {
		result := utils.IsArchiveMimeType(test.mimeType)
		assert.Equal(t, test.expected, result, "IsArchiveMimeType(%s) = %v, expected %v", test.mimeType, result, test.expected)
	}
}

func TestIsCodeFile(t *testing.T) {
	tests := []struct {
		filename string
		expected bool
	}{
		{"test.js", true},
		{"test.ts", true},
		{"test.html", true},
		{"test.css", true},
		{"test.py", true},
		{"test.go", true},
		{"test.txt", false},
		{"test.jpg", false},
		{"noextension", false},
		{"", false},
	}

	for _, test := range tests {
		result := utils.IsCodeFile(test.filename)
		assert.Equal(t, test.expected, result, "IsCodeFile(%s) = %v, expected %v", test.filename, result, test.expected)
	}
}

func TestIsDocumentFile(t *testing.T) {
	tests := []struct {
		filename string
		expected bool
	}{
		{"test.doc", true},
		{"test.docx", true},
		{"test.txt", true},
		{"test.pdf", false}, // PDF is not in document extensions
		{"test.jpg", false},
		{"noextension", false},
		{"", false},
	}

	for _, test := range tests {
		result := utils.IsDocumentFile(test.filename)
		assert.Equal(t, test.expected, result, "IsDocumentFile(%s) = %v, expected %v", test.filename, result, test.expected)
	}
}

func TestParseInt64(t *testing.T) {
	tests := []struct {
		input    string
		expected int64
		hasError bool
	}{
		{"123", 123, false},
		{"0", 0, false},
		{"-123", -123, false},
		{"", 0, false},
		{"abc", 0, true},
		{"123.45", 0, true},
	}

	for _, test := range tests {
		result, err := utils.ParseInt64(test.input)
		if test.hasError {
			assert.Error(t, err, "ParseInt64(%s) should return error", test.input)
		} else {
			assert.NoError(t, err, "ParseInt64(%s) should not return error", test.input)
			assert.Equal(t, test.expected, result, "ParseInt64(%s) = %d, expected %d", test.input, result, test.expected)
		}
	}
}

func TestParseFloat64(t *testing.T) {
	tests := []struct {
		input    string
		expected float64
		hasError bool
	}{
		{"123.45", 123.45, false},
		{"0", 0.0, false},
		{"-123.45", -123.45, false},
		{"", 0.0, false},
		{"abc", 0.0, true},
	}

	for _, test := range tests {
		result, err := utils.ParseFloat64(test.input)
		if test.hasError {
			assert.Error(t, err, "ParseFloat64(%s) should return error", test.input)
		} else {
			assert.NoError(t, err, "ParseFloat64(%s) should not return error", test.input)
			assert.Equal(t, test.expected, result, "ParseFloat64(%s) = %f, expected %f", test.input, result, test.expected)
		}
	}
}

func TestFormatTimestamp(t *testing.T) {
	testTime := time.Date(2023, 1, 15, 10, 30, 45, 0, time.UTC)
	expected := "2023-01-15T10:30:45Z"
	result := utils.FormatTimestamp(testTime)
	assert.Equal(t, expected, result)
}

func TestParseTimestamp(t *testing.T) {
	tests := []struct {
		input    string
		hasError bool
	}{
		{"2023-01-15T10:30:45Z", false},
		{"", false},
		{"invalid", true},
	}

	for _, test := range tests {
		result, err := utils.ParseTimestamp(test.input)
		if test.hasError {
			assert.Error(t, err, "ParseTimestamp(%s) should return error", test.input)
		} else {
			assert.NoError(t, err, "ParseTimestamp(%s) should not return error", test.input)
			if test.input != "" {
				expected, _ := time.Parse(time.RFC3339, test.input)
				assert.Equal(t, expected, result, "ParseTimestamp(%s) = %v, expected %v", test.input, result, expected)
			}
		}
	}
}

func TestTruncateString(t *testing.T) {
	tests := []struct {
		input    string
		maxLen   int
		expected string
	}{
		{"hello", 10, "hello"},
		{"hello world", 5, "he..."},
		{"hello world", 8, "hello..."},
		{"hi", 3, "hi"},
		{"hi", 2, "hi"},
		{"hi", 1, "h"},
		{"", 5, ""},
	}

	for _, test := range tests {
		result := utils.TruncateString(test.input, test.maxLen)
		assert.Equal(t, test.expected, result, "TruncateString(%s, %d) = %s, expected %s", test.input, test.maxLen, result, test.expected)
	}
}

func TestSlugify(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"Hello World", "hello-world"},
		{"Hello   World", "hello-world"},
		{"Hello@World!", "hello-world"},
		{"Hello--World", "hello-world"},
		{"", ""},
		{"a", "a"},
		{"A", "a"},
		{"123", "123"},
		{"hello-world", "hello-world"},
		{"--hello--", "hello"},
	}

	for _, test := range tests {
		result := utils.Slugify(test.input)
		assert.Equal(t, test.expected, result, "Slugify(%s) = %s, expected %s", test.input, result, test.expected)
	}
}

func TestCapitalizeFirst(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"hello", "Hello"},
		{"HELLO", "HELLO"},
		{"hELLO", "HELLO"},
		{"", ""},
		{"a", "A"},
		{"123", "123"},
	}

	for _, test := range tests {
		result := utils.CapitalizeFirst(test.input)
		assert.Equal(t, test.expected, result, "CapitalizeFirst(%s) = %s, expected %s", test.input, result, test.expected)
	}
}

func TestCamelToSnake(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"camelCase", "camel_case"},
		{"CamelCase", "camel_case"},
		{"XMLHttpRequest", "xml_http_request"},
		{"simple", "simple"},
		{"", ""},
		{"A", "a"},
	}

	for _, test := range tests {
		result := utils.CamelToSnake(test.input)
		assert.Equal(t, test.expected, result, "CamelToSnake(%s) = %s, expected %s", test.input, result, test.expected)
	}
}

func TestSnakeToCamel(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"snake_case", "snakeCase"},
		{"another_example", "anotherExample"},
		{"simple", "simple"},
		{"", ""},
		{"a_b_c", "aBC"},
		{"_leading", "Leading"},
		{"trailing_", "trailing"},
	}

	for _, test := range tests {
		result := utils.SnakeToCamel(test.input)
		assert.Equal(t, test.expected, result, "SnakeToCamel(%s) = %s, expected %s", test.input, result, test.expected)
	}
}

File: aegis/frontend/src/test/utils/dataTransformation.test.ts
import {
  formatFileSize,
  getMimeTypeFromExtension,
  isImageMimeType,
  isVideoMimeType,
  isAudioMimeType,
  isArchiveMimeType,
  isCodeFile,
  isDocumentFile,
  parseInt64,
  parseFloat64,
  formatTimestamp,
  parseTimestamp,
  truncateString,
  slugify,
  capitalizeFirst,
  camelToSnake,
  snakeToCamel,
  deepClone,
  safeJsonParse,
  safeJsonStringify,
} from '../../utils/dataTransformation';

/*
describe('Data Transformation Utils', () => {
  describe('formatFileSize', () => {
    it('should format 0 bytes', () => {
      expect(formatFileSize(0)).toBe('0 Bytes');
    });

    it('should format bytes', () => {
      expect(formatFileSize(512)).toBe('512 Bytes');
      expect(formatFileSize(1023)).toBe('1023 Bytes');
    });

    it('should format kilobytes', () => {
      expect(formatFileSize(1024)).toBe('1 KB');
      expect(formatFileSize(1536)).toBe('1.5 KB');
    });

    it('should format megabytes', () => {
      expect(formatFileSize(1048576)).toBe('1 MB');
      expect(formatFileSize(1572864)).toBe('1.5 MB');
    });

    it('should format gigabytes', () => {
      expect(formatFileSize(1073741824)).toBe('1 GB');
    });

    it('should format terabytes', () => {
      expect(formatFileSize(1099511627776)).toBe('1 TB');
    });

    it('should handle large numbers', () => {
      expect(formatFileSize(1125899906842624)).toBe('1024 TB');
    });
  });

  describe('getMimeTypeFromExtension', () => {
    it('should return correct MIME types for known extensions', () => {
      expect(getMimeTypeFromExtension('test.txt')).toBe('text/plain');
      expect(getMimeTypeFromExtension('document.pdf')).toBe('application/pdf');
      expect(getMimeTypeFromExtension('image.jpg')).toBe('image/jpeg');
      expect(getMimeTypeFromExtension('image.jpeg')).toBe('image/jpeg');
      expect(getMimeTypeFromExtension('video.mp4')).toBe('video/mp4');
      expect(getMimeTypeFromExtension('audio.mp3')).toBe('audio/mpeg');
      expect(getMimeTypeFromExtension('archive.zip')).toBe('application/zip');
    });

    it('should return default MIME type for unknown extensions', () => {
      expect(getMimeTypeFromExtension('file.unknown')).toBe('application/octet-stream');
      expect(getMimeTypeFromExtension('file.xyz')).toBe('application/octet-stream');
    });

    it('should handle files without extensions', () => {
      expect(getMimeTypeFromExtension('filename')).toBe('application/octet-stream');
    });

    it('should handle empty strings', () => {
      expect(getMimeTypeFromExtension('')).toBe('application/octet-stream');
    });

    it('should handle case insensitive extensions', () => {
      expect(getMimeTypeFromExtension('test.PDF')).toBe('application/pdf');
      expect(getMimeTypeFromExtension('test.JPG')).toBe('image/jpeg');
    });
  });

  describe('MIME type checks', () => {
    describe('isImageMimeType', () => {
      it('should identify image MIME types', () => {
        expect(isImageMimeType('image/jpeg')).toBe(true);
        expect(isImageMimeType('image/png')).toBe(true);
        expect(isImageMimeType('image/gif')).toBe(true);
      });

      it('should reject non-image MIME types', () => {
        expect(isImageMimeType('text/plain')).toBe(false);
        expect(isImageMimeType('video/mp4')).toBe(false);
        expect(isImageMimeType('application/pdf')).toBe(false);
      });
    });

    describe('isVideoMimeType', () => {
      it('should identify video MIME types', () => {
        expect(isVideoMimeType('video/mp4')).toBe(true);
        expect(isVideoMimeType('video/avi')).toBe(true);
      });

      it('should reject non-video MIME types', () => {
        expect(isVideoMimeType('image/jpeg')).toBe(false);
        expect(isVideoMimeType('text/plain')).toBe(false);
      });
    });

    describe('isAudioMimeType', () => {
      it('should identify audio MIME types', () => {
        expect(isAudioMimeType('audio/mpeg')).toBe(true);
        expect(isAudioMimeType('audio/wav')).toBe(true);
      });

      it('should reject non-audio MIME types', () => {
        expect(isAudioMimeType('image/jpeg')).toBe(false);
        expect(isAudioMimeType('video/mp4')).toBe(false);
      });
    });

    describe('isArchiveMimeType', () => {
      it('should identify archive MIME types', () => {
        expect(isArchiveMimeType('application/zip')).toBe(true);
        expect(isArchiveMimeType('application/x-rar-compressed')).toBe(true);
        expect(isArchiveMimeType('application/x-7z-compressed')).toBe(true);
        expect(isArchiveMimeType('application/gzip')).toBe(true);
      });

      it('should reject non-archive MIME types', () => {
        expect(isArchiveMimeType('image/jpeg')).toBe(false);
        expect(isArchiveMimeType('text/plain')).toBe(false);
      });
    });
  });

  describe('file type checks', () => {
    describe('isCodeFile', () => {
      it('should identify code files', () => {
        expect(isCodeFile('script.js')).toBe(true);
        expect(isCodeFile('component.tsx')).toBe(true);
        expect(isCodeFile('styles.css')).toBe(true);
        expect(isCodeFile('data.json')).toBe(true);
        expect(isCodeFile('program.py')).toBe(true);
        expect(isCodeFile('main.go')).toBe(true);
      });

      it('should reject non-code files', () => {
        expect(isCodeFile('document.txt')).toBe(false);
        expect(isCodeFile('image.jpg')).toBe(false);
        expect(isCodeFile('')).toBe(false);
        expect(isCodeFile('file')).toBe(false);
      });
    });

    describe('isDocumentFile', () => {
      it('should identify document files', () => {
        expect(isDocumentFile('document.doc')).toBe(true);
        expect(isDocumentFile('report.docx')).toBe(true);
        expect(isDocumentFile('notes.txt')).toBe(true);
      });

      it('should reject non-document files', () => {
        expect(isDocumentFile('image.jpg')).toBe(false);
        expect(isDocumentFile('script.js')).toBe(false);
        expect(isDocumentFile('')).toBe(false);
        expect(isDocumentFile('file')).toBe(false);
      });
    });
  });

  describe('parsing functions', () => {
    describe('parseInt64', () => {
      it('should parse valid integers', () => {
        expect(parseInt64('123')).toBe(123);
        expect(parseInt64('0')).toBe(0);
        expect(parseInt64('-456')).toBe(-456);
      });

      it('should return 0 for invalid inputs', () => {
        expect(parseInt64('')).toBe(0);
        expect(parseInt64('   ')).toBe(0);
        expect(parseInt64('abc')).toBe(0);
        expect(parseInt64('123.45')).toBe(0);
      });
    });

    describe('parseFloat64', () => {
      it('should parse valid floats', () => {
        expect(parseFloat64('123.45')).toBe(123.45);
        expect(parseFloat64('0')).toBe(0);
        expect(parseFloat64('-456.78')).toBe(-456.78);
      });

      it('should return 0 for invalid inputs', () => {
        expect(parseFloat64('')).toBe(0);
        expect(parseFloat64('   ')).toBe(0);
        expect(parseFloat64('abc')).toBe(0);
      });
    });
  });

  describe('timestamp functions', () => {
    describe('formatTimestamp', () => {
      it('should format date to ISO string', () => {
        const date = new Date('2023-12-25T10:30:45.000Z');
        expect(formatTimestamp(date)).toBe('2023-12-25T10:30:45.000Z');
      });
    });

    describe('parseTimestamp', () => {
      it('should parse valid ISO strings', () => {
        const result = parseTimestamp('2023-12-25T10:30:45.000Z');
        expect(result).toEqual(new Date('2023-12-25T10:30:45.000Z'));
      });

      it('should return null for invalid inputs', () => {
        expect(parseTimestamp('')).toBeNull();
        expect(parseTimestamp('   ')).toBeNull();
        expect(parseTimestamp('invalid-date')).toBeNull();
        expect(parseTimestamp('2023-13-45')).toBeNull();
      });
    });
  });

  describe('string manipulation', () => {
    describe('truncateString', () => {
      it('should not truncate short strings', () => {
        expect(truncateString('hello', 10)).toBe('hello');
      });

      it('should truncate long strings', () => {
        expect(truncateString('hello world', 8)).toBe('hello...');
        expect(truncateString('very long string', 5)).toBe('ve...');
      });

      it('should handle edge cases', () => {
        expect(truncateString('hi', 1)).toBe('h');
        expect(truncateString('hi', 2)).toBe('hi');
        expect(truncateString('hi', 3)).toBe('hi');
      });
    });

    describe('slugify', () => {
      it('should convert strings to slugs', () => {
        expect(slugify('Hello World')).toBe('hello-world');
        expect(slugify('Hello   World')).toBe('hello-world');
        expect(slugify('Hello@World!')).toBe('hello-world');
        expect(slugify('Hello--World')).toBe('hello-world');
      });

      it('should handle special cases', () => {
        expect(slugify('')).toBe('');
        expect(slugify('a')).toBe('a');
        expect(slugify('A')).toBe('a');
        expect(slugify('--hello--')).toBe('hello');
      });
    });

    describe('capitalizeFirst', () => {
      it('should capitalize first letter', () => {
        expect(capitalizeFirst('hello')).toBe('Hello');
        expect(capitalizeFirst('HELLO')).toBe('HELLO');
        expect(capitalizeFirst('hELLO')).toBe('HELLO');
      });

      it('should handle edge cases', () => {
        expect(capitalizeFirst('')).toBe('');
        expect(capitalizeFirst('a')).toBe('A');
        expect(capitalizeFirst('123')).toBe('123');
      });
    });

    describe('camelToSnake', () => {
      it('should convert camelCase to snake_case', () => {
        expect(camelToSnake('camelCase')).toBe('camel_case');
        expect(camelToSnake('CamelCase')).toBe('Camel_case');
        expect(camelToSnake('XMLHttpRequest')).toBe('XML_http_request');
      });

      it('should handle simple cases', () => {
        expect(camelToSnake('simple')).toBe('simple');
        expect(camelToSnake('')).toBe('');
      });
    });

    describe('snakeToCamel', () => {
      it('should convert snake_case to camelCase', () => {
        expect(snakeToCamel('snake_case')).toBe('snakeCase');
        expect(snakeToCamel('another_example')).toBe('anotherExample');
      });

      it('should handle edge cases', () => {
        expect(snakeToCamel('simple')).toBe('simple');
        expect(snakeToCamel('')).toBe('');
        expect(snakeToCamel('a_b_c')).toBe('aBC');
        expect(snakeToCamel('_leading')).toBe('Leading');
        expect(snakeToCamel('trailing_')).toBe('trailing');
      });
    });
  });

  describe('deepClone', () => {
    it('should clone primitive values', () => {
      expect(deepClone(42)).toBe(42);
      expect(deepClone('hello')).toBe('hello');
      expect(deepClone(true)).toBe(true);
      expect(deepClone(null)).toBe(null);
    });

    it('should clone dates', () => {
      const date = new Date();
      const cloned = deepClone(date);
      expect(cloned).toEqual(date);
      expect(cloned).not.toBe(date); // Different reference
    });

    it('should clone arrays', () => {
      const arr = [1, 2, { a: 3 }];
      const cloned = deepClone(arr);
      expect(cloned).toEqual(arr);
      expect(cloned).not.toBe(arr);
      expect(cloned[2]).not.toBe(arr[2]);
    });

    it('should clone objects', () => {
      const obj = { a: 1, b: { c: 2 } };
      const cloned = deepClone(obj);
      expect(cloned).toEqual(obj);
      expect(cloned).not.toBe(obj);
      expect(cloned.b).not.toBe(obj.b);
    });
  });

  describe('JSON utilities', () => {
    describe('safeJsonParse', () => {
      it('should parse valid JSON', () => {
        const obj = { a: 1, b: 'test' };
        const json = JSON.stringify(obj);
        expect(safeJsonParse(json, {})).toEqual(obj);
      });

      it('should return fallback for invalid JSON', () => {
        const fallback = { default: true };
        expect(safeJsonParse('invalid json', fallback)).toEqual(fallback);
        expect(safeJsonParse('', fallback)).toEqual(fallback);
      });
    });

    describe('safeJsonStringify', () => {
      it('should stringify valid objects', () => {
        const obj = { a: 1, b: 'test' };
        expect(safeJsonStringify(obj)).toBe(JSON.stringify(obj));
      });

      it('should return fallback for invalid objects', () => {
        const circular: any = {};
        circular.self = circular;

        expect(safeJsonStringify(circular)).toBe('{}');
        expect(safeJsonStringify(circular, 'fallback')).toBe('fallback');
      });
    });
  });
});
*/